Add core interfaces

matt0bi opened the pull request:
**What this PR does / why we need it**:

It adds a collection of interfaces that define how a plugin and gordon interacts.

**Which issue(s) this PR fixes** 

N/A

**Special notes for your reviewer**:

None

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

cc @spotify/alf

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,86 @@
+    history_log = Attribute('Log of actions performed on the message')
+
+    def __call__(id, data, phase=None):
+        """Initialize an EventMessage.
```
I can't quite understand from the zope docs why one would choose to use `__call__` instead of `__init__` when declaring an interface. Can you tell me why?

matt0bi replied:
Thanks for catching this. I got [this part of the zope docs](https://docs.zope.org/zope.interface/README.html#declaring-provided-interfaces) all wrong. After messing around with interfaces in ipython, I updated all the interfaces to use `__init__`.

The zope docs show an example in which the `__init__` method to be different from other methods because it's a class method, which means that its definition should be part of a `IFooFactory` interface instead of `IFoo`.

I don't see any downsides of letting `IFoo` define an `__init__` method so I'll go with this more concise approach.

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,86 @@
+            metrics (obj): Optional obj used to emit metrics.
+        """
+
+    async def update_phase(event):
```
(and same for the rest IMO)

matt0bi replied:
I changed it to the more readable `event_message` :+1:

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,86 @@
+from zope.interface import Interface
+
+
+class IEventMessage(Interface):
```
Could we get a docstring for the class?

matt0bi replied:
Good idea! I added a docstring, I went for communicating intent and descriptiveness, do you think it explains the role of EventMessages?

matt0bi commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,86 @@
+        """Initialize an EventMessage.
+
+        Args:
+            id (int): Unique message identifier.
```
Yeah, I think this can wait until v2 so we can tackle this once we get there :). From a quick glance, it looks like different message queues use different id schemes (random, sequential, long, short, etc) so the chance is low, but it's there.

This also made me realize that a message id doesn't have to be an integer, it could be an alphanumeric string, so I changed the type of `msg_id` to `str` (from `int`).

matt0bi commented:
Thanks for checking this out @lynn - I made the requested changes, can you PTAL?

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+            config (dict): plugin-specific configuration
+            success_channel (asyncio.Queue): a sink for successfully created
+                IEventMessages.
+            error_channel (asyncio.Queue): a sink for invalid messages.
```
I think "successfully created IEventMessage" and "invalid messages" is not the correct wording. Not sure what would be correct, but something like "sink for messages when plugin client successfully processes message" and "...failed to process message"

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+        """
+
+
+class IEventClient(IGenericPlugin):
```
Hmm just thought: we have `IEnricherClient` and `IPublisherClient`. Maybe this should be `IEventConsumerClient`?

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+class IEventClient(IGenericPlugin):
+
+    async def start():
+        """Schedule the EventClient task for execution in an event loop.
```
I am thinking you meant to say `EventClient` and not `IEventClient`, and mean `EventClient` to be the implementing class. But it might be confusing. Perhaps something simpler like "Start consuming (processing?) messages in the provided event loop".

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+class IGenericPlugin(Interface):
+    """Do not implement this interface directly.
+
+    Use IEventClient, IEnricherClient, or IPublisherClient instead.
```
Could you wrap each object here in back-ticks? [Quick example](https://github.com/spotify/gordon-gcp/blob/develop/gordon_gcp/compute.py#L17). Sphinx will format it nicely :) I think you can even have them auto-linked if you do something like 

```
:ref:`gordon.interfaces.IEventClient`
```

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+        """
+
+
+class IEnricherClient(IGenericPlugin):
```
Maybe we can get some docstrings on all three clients like the event message :) I really like the tone/approach you did with the event message one.

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,107 @@
+    interface in order to route them to other plugins, and handle retries
+    or cleanup in case of errors.
+    """
+    msg_id = Attribute('Unique message id')
```
another silly zope q: why both an `Attribute` and in the initializer?

matt0bi replied:
I think it makes it clearer that the implementing object is expected to have these attributes vs. just accepting them in the initializer.

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,123 @@
+        """
+
+
+class IEventConsumerClient(IGenericPlugin):
```
I wonder if we should (or even, if there's a way) declare the phase for each plugin, since we're only supporting certain explicitly defined ones (something like `consume` (or `event`?), `enrich`, `publish`, `done`, `drop`). Or, are we trying to be forward-thinking to allow flexibility?

matt0bi replied:
I'm going to go for the "more flexibility" direction and leave this part more open to change. We can revisit this in a while and see if we need to tighten things up instead.

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,123 @@
+
+    async def start():
+        """Begin consuming messages using the provided event loop.
+        """
```
Nitpick - can this be all on one line?

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,123 @@
+
+class IPublisherClient(IGenericPlugin):
+    """Client for publishing processed events to their destination.
+    """
```
Same nitpick here 😬

econchick commented:
@matt0bi in a separate commit (or PR, which ever) could you also bump the dev version & add git tags so that it's released on PyPI?

matt0bi commented:
@econchick I addressed your feedback, can you please take another look?

Also, I can't reply to the comment about `zope.interface.Attribute` so I'll do it here:

> Having class attributes that are actually expected to be specific to each instantiation of the class doesn't make much sense to me, and will actually lead to ugly behavior.
> So, I'd argue all these attributes should actually just be args to the __init__ method, rather than leading to potential confusion that I had (are these meant to be class attributes?).
> I'm open to being convinced otherwise if you feel strongly about Attribute.

I don't feel strongly about it and I can see how that's confusing. I googled around to see if there's a way to specify class/instance attributes with zope but I haven't found anything. The zope docs do say that methods are _attributes_ on a class, which makes sense, but doesn't really help with specifying object attributes, so went with having the attributes be args to the initializer.

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,118 @@
+    or cleanup in case of errors.
+    """
+
+    def __init__(msg_id, data, phase=None, history_log=None):
```
Will there ever be a time where a new event message will be initialized with some content in the history log? I'm wondering if there should be some `setter`/getter functionality instead that the `append_to_history` makes use of (I don't know, just typing half-baked/uncaffeinated ideas).

matt0bi replied:
There might be a case when it's init'ed with eg. `[f'Created msg at {datetime.now()}']`, but I think that the usual case will involve init'ing it with an empty list. I'd prefer this to setter/getter methods to keep the interface leaner, unless you feel strongly about this?

econchick replied:
Ah, hmm. Well first, "...the usual case will involve init'ing it with an empty list" is not true when you have `history_log=None`. Granted, it shouldn't be `history_log=[]` in the args. But it's not necessarily obvious to a plugin that an event message should be initialed with an empty list. 

Yeah I don't know the best approach, so let's leave it at this and change later if we want.

matt0bi replied:
Good point @econchick, I switched the `history_log` arg into being a required param, which should make it stand out more (and the fact that it should be a mutable sequence).

econchick commented on gordon/interfaces.py:
```diff
@@ -0,0 +1,118 @@
+    :py:class:`gordon.interfaces.IEnricherClient`,
+    or :py:class:`gordon.interfaces.IPublisherClient` instead.
+    """
+    phase = Attribute('Plugin phase')
```
Just thinking out loud: I wish there was a way - either thru zope or elsewhere - to have an enum of supported values for `phase`. 🤷‍♀️

matt0bi replied:
I think this is possible through zope's [invariants](https://docs.zope.org/zope.interface/README.html#invariants) - it would be good if this invariant was the only place where these phases are defined (eg. `gordon.interfaces.PHASES`). Do you know if we will/already define them some place else, like the config?

econchick replied:
Constants might be a good idea. The invariant thing might be as well. I'd say let's move on and add/update if we find we need something like this.

matt0bi commented:
Thanks for the patience @econchick! I've addressed your latest feedback, do you mind taking a look?

econchick commented:
@matt0bi :+1:
